### ISSUE TITLE
Option to disable autoscroll for non-synced lyrics

### ISSUE THREAD

**Issue by kamenminkov** [jacquesh/foo_openlyrics#86]
I think this should come by default, but it'd be nice to have a toggle for it. As it is now, if `Scroll type` is set to `Automatic`, even non-synced lyrics are being scrolled, which isn't an issue if the panel is tall enough, but if it's not (i.e. being just tall enough to fit the whole text), it would start off with the start in the middle and gradually scroll to the top, resulting in that not the whole text is visible.

Now that I think of it, it would make sense even for synced lyrics to stay fixed if there is enough height.

**Comment by anemochore** [comment]
+1. Lyric Show 3 provides manual scrolling by drag-and-drop (regarlessly with synced) and we want the feature.

**Comment by jacquesh** [comment]
Just to be sure, the behaviour you want is just for it to not scroll if all the lyrics would fit on the panel at once, is that correct?

What would you want to happen if the lyrics are 1 line too tall (assuming vertical scrolling) for synced lyrics? Would it scroll just one line over the course of the entire track so that the first line is right at the top of the panel at the start of the track when its highlighted, and then the text scrolls by very slowly just enough for the bottom line to be right at the bottom of the panel at the end of the track when its highlighted? Or would it switch to being centered as it is now?

**Comment by kamenminkov** [comment]
To preface the problem, I've only set Scroll type: Automatic because this seems to be the only way to highlight the current line for synced lyrics (setting Scroll type: Manual disables the highlighting as well unfortunately). This should probably go into another issue (unless you're aware of it already).

My main concern is when there's enough height in the pane and yet non-synced lyrics insist on scrolling. Say you have 60 lines of lyrics and the pane is tall enough to contain 70. What happens is that if you enable automatic scrolling, at the beginning of the song there's only the first half of the text visible, and towards the end of the song only the second half is visible. Screenshots to illustrate (with an intentionally small window that barely fits the lyrics - but this also happens with larger windows and longer lyrics) - first one with Scroll type: Automatic, second one - Manual.

![Screenshot 2022-01-20 111752](https://user-images.githubusercontent.com/16286572/150309933-87c9b108-5bb7-4649-b683-cd194934aea5.png)
![Screenshot 2022-01-20 111812](https://user-images.githubusercontent.com/16286572/150309955-201dee10-e894-4e49-8f2d-653b7f880836.png)

**Comment by jacquesh** [comment]
Yeah ok that's what I thought you meant.

Its come up before, I'll look into it. The implementation is complicated a little by the fact that whatever we do needs to make sense in the larger context of other options and its not necessarily clear how this change maps to synced lyrics (as mentioned above). It'd also need to apply to horizontal scrolling, although presumably there can be an analogous change there (don't scroll if you don't need to).

I suppose we could look to LSP3/ESLyric for inspiration. Do you know what they do?

**Comment by kamenminkov** [comment]
> LSP3

Yeah, I used to use it - and I'm pretty happy with its behaviour regarding what we discuss here - but it lacks in automatic searching, sources and so on (it seems to be out of active development, one of the reasons I tried OpenLyrics to begin with).

**Comment by Shamtam** [comment]
LSP3 would still allow manual scrolling "override," even when set to automatic scroll mode.

My preferred behavior would be the following: if scrolling is initiated when in automatic scroll mode, "temporarily" change to manual scroll mode (this would be reset when loading a new track/lyrics, or if requested via e.g. context menu or pressing <kbd>Esc</kbd>). From my memory, this is more or less how LSP3 worked.

**Comment by jacquesh** [comment]
For reference, I'm collecting a list of all the expected behaviours in [this comment on #312](https://github.com/jacquesh/foo_openlyrics/issues/312#issuecomment-2661588683), if anybody here wants to weigh in on that.

**Comment by Athari** [comment]
Rather than disabling scrolling altogether, I think a much better option would be **"Avoid scrolling lyrics offscreen"** (or **"Try to keep lyrics on screen"**).

The idea is simple:
* clamp the top of the lyrics to the top of the panel at the start of the song;
* scroll as usual in the middle of the song;
* clamp the bottom of the lyrics to the bottom of the panel at the end of the song.

(Switching between 1&2 and 2&3 happens when the current algorithm puts lyrics at the relevant panel edge.)

This guarantees that at any point in time:
* as much of the text is visible as possible;
* at least as much as surrounding context is visible as the current algorithm provides.

The reason for this option is simple: to read the lyrics, I don't need the focused line to be in the middle. I can read from the top of the panel just as well.

This works for both synced and unsynced lyrics. It'd probably make sense to have a separate checkbox for synced and unsynced lyrics, because it may actually make sense for synced lyrics to always be centered vertically.
